File: src/tokenwise/render/ansi.ts
/**
 * Minimal ANSI truecolor utilities.
 * Requires a terminal with COLORTERM=truecolor (Windows Terminal, iTerm2,
 * Alacritty, kitty, WezTerm, and most modern terminal emulators).
 * Respects the NO_COLOR env var — callers can check `useColor` before rendering.
 */

const E = '\x1b'

export const useColor =
  !process.env['NO_COLOR'] &&
  process.stdout.isTTY !== false

export const RESET = `${E}[0m`

/** Set 24-bit foreground colour. */
export const fg = (r: number, g: number, b: number): string =>
  `${E}[38;2;${r};${g};${b}m`

/** Set 24-bit background colour. */
export const bg = (r: number, g: number, b: number): string =>
  `${E}[48;2;${r};${g};${b}m`

/** Visible length of a string, stripping all ANSI escape sequences. */
export const vlen = (s: string): number =>
  s.replace(/\x1b\[[0-9;]*m/g, '').length

/** Right-pad a (possibly ANSI-coded) string to `w` visible characters. */
export const padR = (s: string, w: number): string =>
  s + ' '.repeat(Math.max(0, w - vlen(s)))

/** Left-pad a (possibly ANSI-coded) string to `w` visible characters. */
export const padL = (s: string, w: number): string =>
  ' '.repeat(Math.max(0, w - vlen(s))) + s

/** Linearly interpolate two RGB values. */
export const lerpRGB = (
  a: RGB, b: RGB, t: number,
): RGB => [
  Math.round(a[0] + (b[0] - a[0]) * t),
  Math.round(a[1] + (b[1] - a[1]) * t),
  Math.round(a[2] + (b[2] - a[2]) * t),
]

export type RGB = readonly [number, number, number]

/** Shared colour palette (GitHub dark-inspired green accent scheme). */
export const C = {
  textPrimary : [201, 209, 217] as RGB,   // #c9d1d9
  textBright  : [240, 246, 252] as RGB,   // #f0f6fc
  textMuted   : [125, 133, 144] as RGB,   // #7d8590
  textDim     : [ 72,  79,  88] as RGB,   // #484f58
  bgTile      : [ 22,  27,  34] as RGB,   // #161b22 — empty heatmap cell
  track       : [ 28,  35,  41] as RGB,   // #1c2329 — unfilled bar track
  // Green gradient, dim → bright
  green1      : [ 31,  77,  44] as RGB,   // #1f4d2c
  green2      : [ 46, 160,  67] as RGB,   // #2ea043
  green3      : [ 63, 185,  80] as RGB,   // #3fb950
  green4      : [ 86, 211, 100] as RGB,   // #56d364
  green5      : [126, 231, 135] as RGB,   // #7ee787
  // Accents
  blue        : [ 88, 166, 255] as RGB,   // tokens
  purple      : [188, 140, 255] as RGB,   // project bullet 1
  teal        : [138, 212, 255] as RGB,   // project bullet 2
  red         : [200,  60,  60] as RGB,   // negative delta
} as const

File: src/tokenwise/render/ansi.py
"""ANSI truecolor utilities — Python port of ansi.ts."""
from __future__ import annotations

import os
import re
import sys

type RGB = tuple[int, int, int]

RESET = "\x1b[0m"


def use_color() -> bool:
    return not os.environ.get("NO_COLOR") and getattr(sys.stdout, "isatty", lambda: False)()


def fg(r: int, g: int, b: int) -> str:
    return f"\x1b[38;2;{r};{g};{b}m"


def bg(r: int, g: int, b: int) -> str:
    return f"\x1b[48;2;{r};{g};{b}m"


_ANSI_RE = re.compile(r"\x1b\[[0-9;]*m")


def vlen(s: str) -> int:
    """Visible length of a string, stripping all ANSI escape sequences."""
    return len(_ANSI_RE.sub("", s))


def pad_r(s: str, w: int) -> str:
    return s + " " * max(0, w - vlen(s))


def pad_l(s: str, w: int) -> str:
    return " " * max(0, w - vlen(s)) + s


def lerp_rgb(a: RGB, b: RGB, t: float) -> RGB:
    return (
        round(a[0] + (b[0] - a[0]) * t),
        round(a[1] + (b[1] - a[1]) * t),
        round(a[2] + (b[2] - a[2]) * t),
    )


class _Palette:
    text_primary: RGB = (201, 209, 217)
    text_bright:  RGB = (240, 246, 252)
    text_muted:   RGB = (125, 133, 144)
    text_dim:     RGB = ( 72,  79,  88)
    bg_tile:      RGB = ( 22,  27,  34)
    track:        RGB = ( 28,  35,  41)
    green1:       RGB = ( 31,  77,  44)
    green2:       RGB = ( 46, 160,  67)
    green3:       RGB = ( 63, 185,  80)
    green4:       RGB = ( 86, 211, 100)
    green5:       RGB = (126, 231, 135)
    blue:         RGB = ( 88, 166, 255)
    purple:       RGB = (188, 140, 255)
    teal:         RGB = (138, 212, 255)
    red:          RGB = (200,  60,  60)


C = _Palette()
